src/storage/client_storage.rs

```diff
@@ -0,0 +1,282 @@
+        }
+    }
+}
+
+fn get<T: DeserializeOwned>(key: &str) -> Option<T> {
+    #[cfg(not(feature = "ssr"))]
+    {
+        #[cfg(target_arch = "wasm32")]
+        {
+            let s: String = local_storage()?.get_item(key).ok()??;
+            try_serde_from_string(&s)
+        }
+        #[cfg(not(target_arch = "wasm32"))]
+        {
+            let path = LOCATION
+                .get()
+                .expect("Call the set_dir macro before accessing persistant data")
+                .join(key);
+            let s = std::fs::read_to_string(path).ok()?;
+            try_serde_from_string(&s)
+        }
+    }
+    #[cfg(feature = "ssr")]
+    None
+}
+
+pub struct ClientStorage;
+
+impl StorageBacking for ClientStorage {
+    type Key = String;
+
+    fn set<T: Serialize>(key: String, value: &T) {
+        set(key, value);
+    }
+
+    fn get<T: DeserializeOwned>(key: &String) -> Option<T> {
+        get(key)
+    }
+}
+
+/// A persistent storage hook that can be used to store data across application reloads.
+///
+/// Depending on the platform this uses either local storage or a file storage
+#[allow(clippy::needless_return)]
+pub fn use_persistent<T: Serialize + DeserializeOwned + Default + 'static>(
+    cx: &ScopeState,
+    key: impl ToString,
+    init: impl FnOnce() -> T,
+) -> &UsePersistent<T> {
+    let mut init = Some(init);
+    #[cfg(feature = "ssr")]
+    let state = use_ref(cx, || {
+        StorageEntry::<ClientStorage, T>::new(key.to_string(), init.take().unwrap()())
+    });
+    // if hydration is not enabled we can just set the storage
+    #[cfg(all(not(feature = "ssr"), not(feature = "hydrate")))]
+    let state = use_ref(cx, || {
+        StorageEntry::new(
+            key.to_string(),
+            storage_entry::<ClientStorage, T>(key.to_string(), init.take().unwrap()),
+        )
+    });
+    // otherwise render the initial value and then hydrate after the first render
+    #[cfg(all(not(feature = "ssr"), feature = "hydrate"))]
+    let state = {
+        let state = use_ref(cx, || {
+            StorageEntry::<ClientStorage, T>::new(key.to_string(), init.take().unwrap()())
+        });
+        if cx.generation() == 0 {
+            cx.needs_update();
+        }
+        if cx.generation() == 1 {
+            state.set(StorageEntry::new(
+                key.to_string(),
+                storage_entry::<ClientStorage, T>(key.to_string(), init.take().unwrap()),
+            ));
+        }
+
+        state
+    };
+    cx.use_hook(|| UsePersistent {
+        inner: state.clone(),
+    })
+}
+
+/// A persistent storage hook that can be used to store data across application reloads.
+/// The state will be the same for every call to this hook from the same line of code.
+///
+/// Depending on the platform this uses either local storage or a file storage
+#[allow(clippy::needless_return)]
+#[track_caller]
+pub fn use_singleton_persistent<T: Serialize + DeserializeOwned + Default + 'static>(
```

Comment:
This seems a bit risky to me, won't this create a leakage if the application gets reloaded and a new version of the source code gets created? Also this could cause issues if a component is used in multiple places and a singleton state is not desired.

Here's an idea: we store two hashes for each node in the VirtualDom: a location hash, which is a hash of the filename and function name, and a family hash, which is created by hashing the location hash and the parent node's family hash. When storing into the backing storage, we store first into a set using the location hash as the key, and then have the family hash as a secondary key. This allows nodes that share the same source code but are called at different points in the DOM to both store persistent data. If the developer desires a singleton value, they can pass a flag that will use an empty guid instead of the family hash.

On build, we run a preprocessor that generates a set of all the location hashes for nodes that call use_persistent. On the first call to use_persistent, we compare this set with the set of existing storage entries and invalidate any keys that are not present in the set generated by the preprocessor.

Reply:
> This seems a bit risky to me, won't this create a leakage if the application gets reloaded and the structure of the source code changes, say between releases?

Regardless of how we generate the key used to store the value, we do run the risk of "leaking" storage between updates. This could happen because the line changes in the current system, or because the hook order changes in the system you propose. Or in both systems if a call to use_persistant is entirely removed.

Instead of trying to entirely avoid the problem, we could expose a way to clear values that are associated with the current application, or migrate values?

> Also this seems like an unnecessary handicap to limit this feature to just cases where a singleton value is required, such as if a component is reused in multiple nodes but a different value is desired or each instance.

It is definitely a limitation, but it makes the storage somewhat less fragile. The alternative requires differentiating something about the position of the component in the tree (in your case a hash of that position) which means if the position of the component ever changes, the state will break.

For example:
```rust
fn Component(cx: Scope) -> Element {
    render! {
        if rand::random<bool>() {
            rsx! { OtherComponent {} }
        }
        Persistant {}
        Persistant {}
    }
}

fn OtherComponent(cx: Scope) -> Element {
    None
}

fn Persistant(cx: Scope) -> Element {
    // Use persistent here will have a different path based on a random value. We need to know what the dynamic node index of this node is, so that we can provide unique values to each sibling Persistant component
    use_persistant(...);
    None
}
```

The same issue happens with conditional recursion.

> Here's an idea: we store two hashes for each node in the VirtualDom: a location hash, which is a hash of the file path and function name, and a family hash, which is created by hashing the location hash and the parent node's family hash. When storing into the backing storage, we use the location hash as the primary key, and then have the family hash as a secondary key. This allows nodes that share the same source code but are called at different points in the DOM to both store persistent data without clashing. If the developer desires a singleton value, they can pass a flag that will use an empty guid instead of the family hash.

You also need to include the child's index into the key to make the value unique for two duplicate sibling components (the two `Persistant`'s above).

This could be somewhat less fragile, but it introduces new issues where renaming a component breaks storage, and multiple components with the same name in a file (for example in a separate `mod` statement) breaks storage.

> On build, we run a preprocessor that generates a set of all the location hashes for nodes that call use_persistent. On the first call to use_persistent, we compare this set with the set of existing storage entries and drop entries where the primary key is not present in the set generated by the preprocessor.

Recursion, loops, and control flow in general makes this difficult. We would need to generate some range of hashes that the location could be, only really resolving the hash at runtime.



I'm not sure if there is a version of this hook that users will really be able to trust to not loose data. It seems inherently fragile to code changes compared to the normal use_persistant which makes it only useful for situations like caches.

Reply:
Drafting an addition to this to add the above, as well as to subscribe hooks to changes to the underlying storage entry.

Reply:
@ealmloff here's my changes to add subscription. I also did a bit of reorganizing, let me know what you think! https://github.com/ealmloff/dioxus-std/pull/1